tybalt/data_models.py

```diff
@@ -0,0 +1,245 @@
+"""
+tybalt/data_models.py
+2018 Gregory Way
+
+Methods for loading, transforming, and compressing input gene expression data
+
+Usage:
+
+    from tybalt.data_models import DataModel
+
+    dm = DataModel(filename='example_data.tsv')
+
+    # Transform if necessary
+    dm.transform(how='zscore')
+
+    # Compress input data with various features
+    n_comp = 10
+    dm.pca(n_comp)
+    dm.ica(n_comp)
+    dm.nmf(n_comp)
+    dm.nn(n_comp, model='adage')
+    dm.nn(n_comp, model='tybalt')
+    dm.nn(n_comp, model='ctybalt')
+
+    # Extract compressed data from DataModel object
+    # For example,
+    pca_out = dm.pca_df
+
+    # Combine all models into a single dataframe
+    all_df = dm.combine_models()
+"""
+
+import pandas as pd
+
+from sklearn import decomposition
+from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler
+from keras import backend as K
+from keras.utils import to_categorical
+
+from tybalt.models import Tybalt, Adage, cTybalt
+
+
+class DataModel():
+    """
+    Methods for loading and compressing input data
+
+    Usage:
+    from tybalt.data_models import DataModel
+
+    data = DataModel(filename)
+    """
+    def __init__(self, filename=None, df=False, select_columns=False):
+        self.filename = filename
+        if filename is None:
+            self.df = df
+        else:
+            self.df = pd.read_table(self.filename)
+
+        if select_columns:
+            subset_df = self.df.iloc[:, select_columns]
+            other_columns = range(max(select_columns) + 1, self.df.shape[1])
+            self.other_df = self.df.iloc[:, other_columns]
+            self.df = subset_df
+
+    def transform(self, how):
+        self.transformation = how
+        if how == 'zscore':
+            self.transform_fit = StandardScaler().fit(self.df)
+        elif how == 'zeroone':
+            self.transform_fit = MinMaxScaler().fit(self.df)
+        else:
+            raise ValueError('how must be either "zscore" or "zeroone".')
+
+        self.df = pd.DataFrame(self.transform_fit.transform(self.df),
+                               index=self.df.index,
+                               columns=self.df.columns)
+
+    def pca(self, n_components, transform_df=False):
+        self.pca_fit = decomposition.PCA(n_components=n_components)
+        self.pca_fit = self.pca_fit.fit(self.df)
+        self.pca_df = self.pca_fit.transform(self.df)
```

Comment:
Might want to switch to `fit_transform`.

Reply:
I am interested in retaining the ability to transform new data here, hence the argument to `transform_df`

Reply:
Right. I'm saying change to this `self.pca_df = self.pca_fit.fit_transform(df)`. `self.pca_fit` will be trained and you can transform as you please below.

Reply:
ah cool, I didn't realize `pca.fit_transform()` also retained the ability to `fit()` new data! It looks like this [line ensures that it can transform](https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/decomposition/pca.py#L755).

I also checked time complexity, and the one line `fit_transform()` is slightly faster:

```python
from sklearn import decomposition
import numpy

n_components=10
df = numpy.random.rand(8000,2000)
df2 = numpy.random.rand(2000,2000)

%%time
pca_fit = decomposition.PCA(n_components=n_components)
pca_fit = pca_fit.fit(df)
pca_df = pca_fit.transform(df)

pca_df2 = pca_fit.transform(df2)
# CPU times: user 2.38 s, sys: 356 ms, total: 2.73 s
# Wall time: 778 ms

%%time
pca_fit = decomposition.PCA(n_components=n_components)
pca_df = pca_fit.fit_transform(df)

pca_df2_alt = pca_fit.transform(df2)
# CPU times: user 2.19 s, sys: 304 ms, total: 2.49 s
# Wall time: 717 ms
```